inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,368 @@
+        """Test that update raises the correct exception when an error occurs."""
+        with patch('inbm_common_lib.utility.create_file_with_contents', side_effect=exception):
+            with pytest.raises(expected_exception):
+                manager.update(parameters)
+
+    def test_update_successful(self, manager, parameters, tmp_path):
+        """Test that update successfully updates the source file."""
+        with patch('os.path.join', return_value=str(tmp_path / 'test')), \
+             patch('inbm_common_lib.utility.create_file_with_contents', return_value=None) as mock_create_file:
+            manager.update(parameters)
+            mock_create_file.assert_called_once_with(
+                str(tmp_path / 'test'), ['deb http://archive.ubuntu.com/ubuntu/ bionic main restricted']
+            )
+
+#DO NOT DELETE THIS LINE - TestUbuntuOsSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuOsSourceManagerAdd:
+    @pytest.fixture
+    def manager(self):
+        return UbuntuOsSourceManager()
+
+    @pytest.mark.parametrize('sources', [
+        (['deb http://archive.ubuntu.com/ubuntu/ bionic main restricted']),  # single source
+        (['deb http://archive.ubuntu.com/ubuntu/ bionic main restricted', 'deb http://archive.ubuntu.com/ubuntu/ bionic universe']),  # multiple sources
+        (['deb http://archive.ubuntu.com/ubuntu/ bionic main restricted', 'deb http://archive.ubuntu.com/ubuntu/ bionic main restricted']),  # duplicate sources
+    ])
+    def test_add_valid_sources(self, manager, sources):
+        # Mock the open function
+        with patch('builtins.open', mock_open()) as m:
+            params = SourceParameters(sources=sources)
+            manager.add(params)
+            m.assert_called_once_with('/etc/apt/sources.list', 'a')
+            handle = m()
+            for source in sources:
+                handle.write.assert_any_call(f"{source}\n")
+
+    @pytest.mark.parametrize('sources', [
+        (['http://archive.ubuntu.com/ubuntu/ bionic main restricted']),  # missing deb
+        (['deb archive.ubuntu.com/ubuntu/ bionic main restricted']),  # missing http://
+        (['deb http://archive.ubuntu.com/ubuntu/ bionic']),  # missing main restricted
+    ])
+    def test_add_invalid_sources(self, manager, sources):
+        params = SourceParameters(sources=sources)
+        with pytest.raises(SourceError):
+            manager.add(params)
+
+    def test_add_os_error(self, manager):
+        # Mock the open function to raise an OSError
+        with patch('builtins.open', mock_open()) as m:
+            m.side_effect = OSError
+            params = SourceParameters(sources=['deb http://archive.ubuntu.com/ubuntu/ bionic main restricted'])
+            with pytest.raises(SourceError):
+                manager.add(params)
+
+#DO NOT DELETE THIS LINE - TestUbuntuOsSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuOsSourceManagerList:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)

<h3>Class: UbuntuOsSourceManager |Function: list</h3>

[Source Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py?plain=1#L50)

[Unit Test Case Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py?plain=1#L208)

<h4>Test Case Scenarios:</h4>
Here are the test cases generated by the unit test:

1. Test Case: `test_list(file_content="deb http://archive.ubuntu.com/ubuntu/ bionic main restricted\n", expected_output=["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted"])`
    - Explanation: This test case checks the normal scenario where the source file contains one source line. The expected output is a list containing the source line.

2. Test Case: `test_list(file_content="deb http://archive.ubuntu.com/ubuntu/ bionic main restricted\ndeb-src http://archive.ubuntu.com/ubuntu/ bionic main restricted\n", expected_output=["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted", "deb-src http://archive.ubuntu.com/ubuntu/ bionic main restricted"])`
    - Explanation: This test case checks the normal scenario where the source file contains multiple source lines. The expected output is a list containing the source lines.

3. Test Case: `test_list(file_content="# This is a comment\n", expected_output=[])`
    - Explanation: This test case checks the edge case scenario where the source file does not contain any valid source lines, only a comment. The expected output is an empty list.

4. Test Case: `test_list(file_content="invalid line\n", expected_output=[])`
    - Explanation: This test case checks the edge case scenario where the source file contains an invalid source line. The expected output is an empty list.

5. Test Case: `test_list(file_content="deb http://archive.ubuntu.com/ubuntu/ bionic main restricted\ninvalid line\n", expected_output=["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted"])`
    - Explanation: This test case checks the normal scenario where the source file contains both valid and invalid source lines. The expected output is a list containing only the valid source line.

6. Test Case: `test_list_error(file_content=None, exception_message="Error opening source file: [Errno 2] No such file or directory: '/etc/apt/sources.list'")`
    - Explanation: This test case checks the error scenario where the source file does not exist. The expected output is a `SourceError` exception with a specific error message.

These test cases cover a wide range of possible inputs and scenarios for the `list` function of the `UbuntuOsSourceManager` class. They ensure that the function behaves correctly in normal, edge case, and error scenarios.


<h4>Test Execution Results:</h4>

| Total Cases    | 6  |
| -------------- | -- |
| Passed Cases   | 6  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)